compiler/rustc_data_structures/src/sso/set.rs

```diff
@@ -0,0 +1,342 @@
+use super::EitherIter;
+use crate::fx::FxHashSet;
+use arrayvec::ArrayVec;
+use std::borrow::Borrow;
+use std::fmt;
+use std::hash::Hash;
+use std::iter::FromIterator;
+
+/// Small-storage-optimized implementation of a set.
+///
+/// Stores elements in a small array up to a certain length
+/// and switches to `HashSet` when that length is exceeded.
+///
+/// Implements subset of HashSet API.
+///
+/// Missing HashSet API:
+///   all hasher-related
+///   try_reserve (unstable)
+///   shrink_to (unstable)
+///   drain_filter (unstable)
+///   get_or_insert/get_or_insert_owned/get_or_insert_with (unstable)
+///   difference/symmetric_difference/intersection/union
+///   is_disjoint/is_subset/is_superset
+///   PartialEq/Eq (requires sorting the array)
+///   BitOr/BitAnd/BitXor/Sub
+#[derive(Clone)]
+pub enum SsoHashSet<T> {
```

Comment:
Why is this not just a wrapper struct around `SsoHashMap<T, ()>`?

Reply:
That's actually a good idea.
Sure, it's already a thin wrap, but it could be even thinner.
Let me try that.

Reply:
Ah :laughing: I thought there was a reason and just wanted it documented. Maybe let's do that in a separate step, since we already have a perf situation